packages/har-remix/src/getCertificateInfo.ts

```diff
@@ -0,0 +1,121 @@
+import * as fs from 'fs';
+import { asn1, md, pki, util } from 'node-forge';
+import { homedir } from 'os';
+import * as path from 'path';
+import { CertificateInfo } from '../types';
+
+export default function getCertificateInfo(
+  cacheDir = getCacheDir()
+): CertificateInfo {
+  let certInfo = readCertificateInfo(cacheDir);
+  if (certInfo === undefined) {
+    certInfo = createCertificateAndPrivateKey();
+    saveCertificateInfo(certInfo, cacheDir);
+  }
+  return certInfo;
+}
+
+function getCacheDir(dir = homedir()) {
+  return path.join(dir, '.har-remix');
+}
+
+function readCertificateInfo(
+  cacheDir = getCacheDir()
+): CertificateInfo | undefined {
+  const text = tryRead(path.join(cacheDir, 'certificate.json'));
+  if (text !== undefined) {
+    return validateCertificateInfo(text);
+  }
+}
+
+function validateCertificateInfo(text: string): CertificateInfo | undefined {
+  try {
+    const info: CertificateInfo = JSON.parse(text);
+    const cert = pki.certificateFromPem(info.cert);
+    if (
+      Date.now() < cert.validity.notBefore.getTime() ||
+      Date.now() > cert.validity.notAfter.getTime()
+    ) {
+      return;
+    }
+    pki.privateKeyFromPem(info.key);
+    const fingerprint = util.decode64(info.fingerprint);
+    if (fingerprint.length !== 32) {
+      return;
+    }
+    return info;
+  } catch (e) {
+    // TODO use debug
+    // tslint:disable-next-line: no-console
+    console.error(`cached certificate is invalid, regenerating`);
+  }
+}
+
+function saveCertificateInfo(
+  certificateInfo: CertificateInfo,
+  cacheDir = getCacheDir()
+) {
+  ensureDir(cacheDir);
+  fs.writeFileSync(
+    path.join(cacheDir, 'certificate.json'),
+    JSON.stringify(certificateInfo)
+  );
+}
+
+function createCertificateAndPrivateKey(): CertificateInfo {
+  const pair = pki.rsa.generateKeyPair(2048);
+  const cert = pki.createCertificate();
+  cert.setSubject([
+    { shortName: 'CN', value: 'localhost' },
+    { shortName: 'OU', value: 'Test' },
+    { shortName: 'O', value: 'Test' },
+    { shortName: 'L', value: 'Sunnyvale' },
+    { shortName: 'ST', value: 'California' },
+    { shortName: 'C', value: 'US' },
+  ]);
+  cert.publicKey = pair.publicKey;
+  cert.serialNumber = '01';
+
+  const notBefore = new Date();
+  const notAfter = new Date();
+  notAfter.setFullYear(notBefore.getFullYear() + 5);
```

Comment:
Have you thought about how long you want this certificate to last? When I've had students create certs for workshops, I try to have them expire rather quickly to limit the impact of potential bugs (i.e., if someone did something that made trusting the cert dangerous, like checking their private key into git).

Having a compromised user-trusted root can be _really_ dangerous, as protections like [public key pinning](https://developer.mozilla.org/en-US/docs/Web/HTTP/Public_Key_Pinning) are ignored if an attacker were to use the self-signed root, and [HTTP strict-transport security](https://tools.ietf.org/html/rfc6797#page-37) can be abused to trick users into thinking an attacker's site is genuine, and the genuine site is dangerous.

Reply:
This isn't user trusted and will not be asking the user to ignore it. It will be to launch chrome with a temp user data dir (not the user's) and a flag to ignore cert errors from this certificate.

Reply:
it is like this https://github.com/catapult-project/catapult/commit/724bb776bcdea642a5bc9cc20fe14bb0282b4900#diff-b00b713a587ac8914d901ceebfc0dd02R154

This flag does not work unless chrome has a custom --user-data-dir flag too which in our case points to a fresh empty directory.

Unfortunately this doesn't work with headless, but we can still use the dev tools api to ignore the cert.

If for some reason we have to trust these, I will revisit the expiration.  I guess another danger would be the user adds it to their keychain as part trying various things during troubleshooting.